### DISCUSSION TITLE
How to add bearer token that may need updating to auth header?

### DISCUSSION THREAD

**Discussion by faph** [aio-libs/aiohttp#6908] [Q&A] [answered]
Trying to figure out how best passing a bearer token with an aiohttp client when sending requests.

I understand i can pass `headers` to both the client and individual request methods, but the problem is that thesea are "fixed" headers. The bearer token itself needs periodic updating and such an update may need to take place at any request being made.

What's the best approach to configure a client with such a "dynamic" auth header (without having to evaluate and pass the header manually prior to making every request)?

This is basically the same question as https://stackoverflow.com/questions/63248239/automatically-refresh-auth-token-in-aiohttp

**Comment by Dreamsorcerer** [comment, marked as answer]
Just a quick guess from looking at the documentation. The headers appear to be accessible on the ClientSession: https://docs.aiohttp.org/en/stable/client_reference.html#aiohttp.ClientSession.headers

So, just update the header value there?

**Reply by faph** [reply]
Understood. But that would require that header to be explicitly set/updated **prior** to making each request. 

I’m really looking for an option where the client is configured once and then each request Will automatically trigger a header update if and when required.

**Reply by Dreamsorcerer** [reply]
How would it know when it is required or how to update the header?

It's just an HTTP client, how is it going to know about your application logic?

**Reply by faph** [reply]
For example by configuring the client with a callable or similar. As in, a callable I write. 

References:

- https://requests.readthedocs.io/en/latest/user/authentication/#new-forms-of-authentication
- https://www.python-httpx.org/advanced/#customizing-authentication

**Reply by Dreamsorcerer** [reply]
Yep, interesting. I don't think we support anything like that currently.
I also feel like those examples are pushing too much application logic into the library (the requests one seems a little light on how it actually works, while the httpx one looks rather complicated with a whole bunch of different things being possible). You could just create a custom class for you application requests for now.

e.g. Roughly something like:
```
class MyClient:
   def __init__(self):
       self._client = ClientSession()
       self._url = "https://my.domain"

    async def request(self, endpoint, method="GET"):
        async with self._client.request(self._url + endpoint, method=method) as resp:
             if resp.status == 401:
                 # Do auth and retry
            return await resp.json()

    async def __aenter__(self):
        await self._client.__aenter__()

    async def __aexit__(self):
        await self._client.__aexit__()
```

I'm not too sure whether I prefer something like this or those auth classes.
If you'd like to work on adding such a feature though, then maybe create a proposal for how the API would look, and once agreed, a PR to implement it.

**Reply by faph** [reply]
Cool. Let me think about a proposal.

**Reply by faph** [reply]
Proposal drafted here: #6915
